modules/library/main/src/main/java/org/geotools/data/sort/SimpleFeatureIO.java

```diff
@@ -269,6 +276,12 @@ private byte[] getByteBuffer(int length) {
 
     @SuppressWarnings("BanSerializableRead")
     private Object readObject() throws IOException {
+        if ("false".equalsIgnoreCase(this.enableDeserialization)) {
+            throw new IllegalStateException("Object deserialization is not allowed");
+        } else if (!this.initialFileEmpty && !"true".equalsIgnoreCase(this.enableDeserialization)) {
+            throw new IllegalStateException(
+                    "Object deserialization is only allowed when created with an empty file");
+        }
```

Comment:
SimpleFeatureIO is used to perform sort/merge over secondary memory (when past a certain limit of features, below it works in memory). This is the default behavior of ContentDataStore, so all stores inherit it unless they can provide native sorting. 

Now, most attributes will be simple types and won't be affected, but it's not uncommon for stores nowadays to add Map objects as attributes to support more complex types. For example, PostgreSQL support for hstore and JSON uses maps (likely saved by native order by encoding), and I'm guessing ElasticSearch and Mongo too (but they also support native sorting).

Long story short, while I cannot find an example right away, the chance of breaking some existing implementation out there is not insignificant. This should probably be marked in the upgrade notes for both GeoTools and GeoServer.

Reply:
There are two existing cases where SimpleFeatureIO is used and both create a single SimpleFeatureIO object on an empty temporary file and then write to and read from the file from the same object. The default behavior was specifically chosen to allow this behavior.
https://github.com/geotools/geotools/blob/main/modules/library/main/src/main/java/org/geotools/data/sort/MergeSortDumper.java#L137
https://github.com/geotools/geotools/blob/main/modules/library/render/src/main/java/org/geotools/renderer/lite/MarkFeatureIterator.java#L161

Reply:
Isn't "enableDeserialization" is false by default? Then the second if won't be reached at all... Oh... by default it's the empty string... that's why I found it confusing, normally for boolean system variables we use Boolean.getBoolean instead.